### ISSUE TITLE
Are all keras layers supported for weight pruning?

### ISSUE THREAD

**Issue by anuar12** [tensorflow/model-optimization#39]
Hi, thanks for a very convenient package!
Are the layers like LSTM, ConvLSTM2D and TimeDistributed supported for weight pruning? In the other issue, someone had a problem with TimeDistributed layer.
If they are not supported out of the box, would inheriting from PrunableLayer and implementing `get_prunable_weights()` (perhaps as return empty list) fix the issue?

Thanks!

**Comment by s36srini** [comment]
Based on the _update_mask function in pruning_impl.py, any weight tensor should be prunable as it simply sorts the weights and updates the mask based on the smallest value. Don't see why this logic should be restricted to a given layer.

**Comment by alanchiao** [comment]
HI Anuar,

We should be supporting all builtin-Keras layers out-of-the-box, though in the implementation, we need to add them one-by-one (see [this](https://github.com/tensorflow/model-optimization/blob/master/tensorflow_model_optimization/python/core/sparsity/keras/prune_registry.py)). Keras wrappers are a bit different, and in as you can see in the [other issue](https://github.com/tensorflow/model-optimization/issues/6), they aren't supported yet (BiDirectional and TimeDistributed).

As you suggested, anything missing (either a custom Keras layer ... or something that hasn't been added above yet) can inherit from PrunableLayer (as suggested [here](https://www.tensorflow.org/model_optimization/guide/pruning/train_sparse_models#prune_a_custom_layer)).

What Sanjay said is correct also.

**Comment by bothrasumit** [comment]
@alanchiao 
Is pruning supported for a model with BiDirectional LSTM layer and TimeDistributed Dense Layer?

**Comment by alanchiao** [comment]
@bothrasumit: not yet. See [this](https://github.com/tensorflow/model-optimization/blob/master/tensorflow_model_optimization/python/core/sparsity/keras/prune_registry.py) for currently supported layers.

**Comment by oholimoli** [comment]
I'm trying to prune some SimpleRNN layers of my model.

Are RNN layers now supported? If not, how can I make a custom prunable RNN layer?

I followed the instructions in [custom_training_loop](https://www.tensorflow.org/model_optimization/guide/pruning/comprehensive_guide#custom_training_loop) and [prune_some_layers](https://www.tensorflow.org/model_optimization/guide/pruning/comprehensive_guide#prune_some_layers_sequential_and_functional). When I fit my model, I don't get an error but the weights are not pruned :(.

In the [prune_registry](https://github.com/tensorflow/model-optimization/blob/master/tensorflow_model_optimization/python/core/sparsity/keras/prune_registry.py) the SimpleRNN layer is mentioned here:

```
  _RNN_LAYERS = frozenset({
      layers.GRU,
      layers.LSTM,
      layers.RNN,
      layers.SimpleRNN,
  })
```
